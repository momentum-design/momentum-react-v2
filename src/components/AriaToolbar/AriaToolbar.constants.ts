const CLASS_PREFIX = 'md-aria-toolbar';

const DEFAULTS = {
  ORIENTATION: 'horizontal' as const,
  SHOULD_RENDER_AS_BUTTON_GROUP: false,
};

const STYLE = {
  wrapper: `${CLASS_PREFIX}-wrapper`,
};

export { CLASS_PREFIX, DEFAULTS, STYLE };
