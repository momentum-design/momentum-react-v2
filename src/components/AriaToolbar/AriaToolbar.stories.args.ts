import { commonStyles } from '../../storybook/helper.stories.argtypes';

const ariaToolbarArgTypes = {
  children: {
    description: 'Provides the child nodes for this element.',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'ReactNode',
      },
      defaultValue: {
        summary: 'undefined',
      },
    },
  },
  ariaLabel: {
    description: 'The aria-label for the toolbar',
    control: { type: 'text' },
    table: {
      type: {
        summary: 'string',
      },
      defaultValue: {
        summary: 'aria toolbar',
      },
    },
  },
};

export { ariaToolbarArgTypes };

export default {
  ...commonStyles,
  ...ariaToolbarArgTypes,
};
