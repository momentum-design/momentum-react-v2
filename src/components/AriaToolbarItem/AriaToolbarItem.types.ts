import { CSSProperties, ReactElement, ReactNode } from 'react';

export interface Props {
  /**
   * Child components of this AriaToolbarItem.
   */
  children: ReactElement;

  /**
   * The index of this item in the toolbar. Used for focus.
   */
  itemIndex: number;
}
