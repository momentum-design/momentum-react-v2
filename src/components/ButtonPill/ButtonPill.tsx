import React, { forwardRef, RefObject } from 'react';
import classnames from 'classnames';
import ButtonSimple from '../ButtonSimple';

import { DEFAULTS, STYLE } from './ButtonPill.constants';
import { Props } from './ButtonPill.types';
import './ButtonPill.style.scss';

const ButtonPill = forwardRef((props: Props, providedRef: RefObject<HTMLButtonElement>) => {
  const {
    className,
    color,
    disabled,
    shallowDisabled,
    ghost,
    grown,
    outline,
    size,
    inverted,
    onlyTriggersTooltip,
    ...otherProps
  } = props;

  if (ghost && inverted) {
    console.warn('MRV2: Momentum does not support a ghost inverted ButtonPill.');
  }

  if (onlyTriggersTooltip && (outline || inverted || color || disabled || shallowDisabled)) {
    console.warn(
      'MRV2: Momentum does not support a button that onlyTriggersTooltip and is outline, inverted, has color, or is disabled'
    );
  }

  return (
    <ButtonSimple
      className={classnames(STYLE.wrapper, className)}
      ref={providedRef}
      data-color={color || DEFAULTS.COLOR}
      data-disabled={disabled || DEFAULTS.DISABLED}
      data-shallow-disabled={shallowDisabled || DEFAULTS.SHALLOW_DISABLED}
      aria-disabled={shallowDisabled || DEFAULTS.SHALLOW_DISABLED}
      data-ghost={ghost || onlyTriggersTooltip || DEFAULTS.GHOST}
      data-grown={grown || DEFAULTS.GROWN}
      data-outline={outline || DEFAULTS.OUTLINE}
      data-size={size || DEFAULTS.SIZE}
      data-inverted={inverted || DEFAULTS.INVERTED}
      data-only-triggers-tooltip={onlyTriggersTooltip || DEFAULTS.ONLY_TRIGGERS_TOOLTIP}
      isDisabled={disabled}
      {...otherProps}
    />
  );
});

ButtonPill.displayName = 'ButtonPill';

export default ButtonPill;
