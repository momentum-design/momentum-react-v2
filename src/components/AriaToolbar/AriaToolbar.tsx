import React, { FC, useRef, useState, useEffect, useCallback } from 'react';
import classnames from 'classnames';

import { DEFAULTS, STYLE } from './AriaToolbar.constants';
import { Props } from './AriaToolbar.types';
import ButtonGroup from '../ButtonGroup';
import { AriaToolbarContext } from './AriaToolbar.utils';
import { getKeyboardFocusableElements } from '../../utils/navigation';

/**
 * The AriaToolbar component. A style-less by default or button-group styled component implementing the Aria Toolbar pattern
 * see https://www.w3.org/WAI/ARIA/apg/patterns/toolbar
 */
const AriaToolbar: FC<Props> = (props: Props) => {
  const {
    ariaLabel,
    className,
    id,
    style,
    children,
    orientation = DEFAULTS.ORIENTATION,
    shouldRenderAsButtonGroup = DEFAULTS.SHOULD_RENDER_AS_BUTTON_GROUP,
    onTabPress,
    ariaControls,
    buttonGroupProps,
    ariaToolbarItemsSize,
    ...rest
  } = props;

  const [currentFocus, setCurrentFocus] = useState(undefined);

  const buttonRefs = useRef({});
  const ref = useRef<HTMLDivElement>(null);

  useEffect(() => {
    buttonRefs.current[currentFocus]?.focus();
  }, [currentFocus]);

  const getContext = useCallback(
    () => ({
      currentFocus,
      setCurrentFocus,
      orientation,
      onTabPress,
      ariaToolbarItemsSize,
      buttonRefs,
    }),
    [currentFocus, setCurrentFocus, orientation, onTabPress, ariaToolbarItemsSize, buttonRefs]
  );

  const commonProps = {
    className: classnames(className, STYLE.wrapper),
    id: id,
    style: style,
    'aria-label': ariaLabel,
    'aria-controls': ariaControls,
    role: 'toolbar',
    'aria-orientation': orientation,
    ...rest,
  };

  const renderBody = () => {
    if (shouldRenderAsButtonGroup) {
      return (
        <ButtonGroup ref={ref} {...buttonGroupProps} {...commonProps} orientation={orientation}>
          {children}
        </ButtonGroup>
      );
    } else {
      return (
        <div ref={ref} {...commonProps}>
          {children}
        </div>
      );
    }
  };

  useEffect(() => {
    // When the toolbar is rendered inside a list, only the first item in the toolbar
    // should be focusable. This is to preserve the tab order as the
    // List uses a roving tab index.
    getKeyboardFocusableElements(ref.current, false).forEach((el, index) => {
      if (index === 0) {
        return;
      }
      el.setAttribute('data-exclude-focus', 'true');
    });
  }, [ref]);

  return (
    <AriaToolbarContext.Provider value={getContext()}>{renderBody()}</AriaToolbarContext.Provider>
  );
};

export default AriaToolbar;
