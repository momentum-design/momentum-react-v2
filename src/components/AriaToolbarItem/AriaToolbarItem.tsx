import { Props } from './AriaToolbarItem.types';

import React, { forwardRef, useCallback } from 'react';

import { useKeyboard } from '@react-aria/interactions';
import { useAriaToolbarContext } from '../AriaToolbar/AriaToolbar.utils';

const AriaToolbarItem = forwardRef<HTMLButtonElement, Props>((props, providedRef) => {
  const { children, itemIndex } = props;

  const {
    currentFocus,
    setCurrentFocus,
    buttonRefs,
    orientation,
    onTabPress,
    ariaToolbarItemsSize,
  } = useAriaToolbarContext();

  const { keyboardProps } = useKeyboard({
    onKeyDown: (e) => {
      // for the escape key (and other key presses), continue propagation to let Popovers / Modals know that
      // they should close
      e.continuePropagation();

      switch (e.key) {
        case orientation === 'horizontal' ? 'ArrowLeft' : 'ArrowUp':
          e.preventDefault();
          setCurrentFocus((ariaToolbarItemsSize + (currentFocus || 0) - 1) % ariaToolbarItemsSize);
          break;

        case orientation === 'horizontal' ? 'ArrowRight' : 'ArrowDown':
          e.preventDefault();
          setCurrentFocus((ariaToolbarItemsSize + (currentFocus || 0) + 1) % ariaToolbarItemsSize);
          break;

        case 'Tab': {
          if (onTabPress) {
            onTabPress(e);
          }
          break;
        }

        default:
          break;
      }
    },
  });

  const getPropsForChildren = useCallback(
    (child, index) => {
      return {
        tabIndex: index === (currentFocus || 0) ? 0 : -1,
        ref: (e: HTMLButtonElement) => {
          buttonRefs.current[index] = e;
          if (providedRef) {
            if (typeof providedRef === 'function') {
              providedRef(e);
            }
          }
        },
        onFocus: (e) => {
          setCurrentFocus(index);
          if (child.props.onFocus) {
            child.props.onFocus(e);
          }
        },
        onPress: () => {
          setCurrentFocus(index);
          if (child.props.onPress) {
            child.props.onPress();
          }
        },
        useNativeKeyDown: true,
        ...keyboardProps,
      };
    },
    [currentFocus]
  );

  return React.cloneElement(children, getPropsForChildren(children, itemIndex));
});

export default AriaToolbarItem;
