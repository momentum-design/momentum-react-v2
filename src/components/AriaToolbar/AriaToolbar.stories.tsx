import React from 'react';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import AriaToolbar from './';
import Documentation from './AriaToolbar.stories.docs.mdx';
import ButtonCircle from '../ButtonCircle';
import ButtonPill from '../ButtonPill';

import argTypes from './AriaToolbar.stories.args';

export default {
  title: 'Momentum UI/AriaToolbar',
  component: AriaToolbar,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

const onPressHandler = () => {
  const input: HTMLInputElement = document.getElementById('textInput') as HTMLInputElement;
  input.value = input.value + 'a';
};

const Horizontal = (props) => {
  return (
    <>
      <AriaToolbar
        ariaLabel="toolbar 1"
        ariaControls="textInput"
        style={{ display: 'flex', columnGap: '0.5rem', marginBottom: '1rem' }}
        {...props}
      >
        <ButtonPill onPress={onPressHandler}>Item 1</ButtonPill>
        <ButtonCircle onPress={onPressHandler}>2</ButtonCircle>
        <ButtonCircle onPress={onPressHandler}>3</ButtonCircle>
      </AriaToolbar>
      <input type="text" id="textInput" aria-label="A text input" />
    </>
  );
};

Horizontal.argTypes = argTypes;

const Vertical = () => {
  return (
    <>
      <AriaToolbar
        orientation="vertical"
        ariaControls="textInput"
        ariaLabel="toolbar 1"
        style={{ display: 'flex', rowGap: '0.5rem', flexDirection: 'column' }}
      >
        <ButtonCircle onPress={onPressHandler}>1</ButtonCircle>
        <ButtonCircle onPress={onPressHandler}>2</ButtonCircle>
        <ButtonCircle onPress={onPressHandler}>3</ButtonCircle>
      </AriaToolbar>
      <AriaToolbar
        orientation="vertical"
        ariaControls="textInput"
        ariaLabel="toolbar 2"
        style={{
          display: 'flex',
          rowGap: '0.5rem',
          flexDirection: 'column',
          marginTop: '1rem',
          marginBottom: '1rem',
        }}
      >
        <ButtonCircle onPress={onPressHandler}>1</ButtonCircle>
        <ButtonCircle onPress={onPressHandler}>2</ButtonCircle>
        <ButtonCircle onPress={onPressHandler}>3</ButtonCircle>
      </AriaToolbar>
      <input type="text" id="textInput" aria-label="A text input" />
    </>
  );
};

export { Horizontal, Vertical };
