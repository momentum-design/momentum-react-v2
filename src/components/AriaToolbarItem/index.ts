import { default as AriaToolbarItem } from './AriaToolbarItem';
import { Props } from './AriaToolbarItem.types';

export type AriaToolbarItemProps = Props;

export default AriaToolbarItem;
