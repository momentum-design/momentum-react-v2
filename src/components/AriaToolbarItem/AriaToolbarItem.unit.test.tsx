import React from 'react';
import { mount } from 'enzyme';

import AriaToolbarItem from './';
import ButtonPill from '../ButtonPill';
import { AriaToolbarContext } from '../AriaToolbar/AriaToolbar.utils';

const TestComponent = (props) => {
  return <AriaToolbarItem itemIndex={0} {...props} children={<ButtonPill>Test</ButtonPill>} />;
};

const mountWithContext = (component) =>
  mount(component, {
    wrappingComponent: AriaToolbarContext.Provider,
    wrappingComponentProps: {
      value: {
        currentFocus: 0,
        setCurrentFocus: jest.fn(),
        buttonRefs: { current: [{ focus: jest.fn() }] },
        orientation: 'horizontal',
        onTabPress: jest.fn(),
        ariaToolbarItemsSize: 1,
      },
    },
  });

describe('<AriaToolbarItem />', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mountWithContext(<TestComponent />);

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('passes extra props to children', () => {
      const container = mountWithContext(
        <AriaToolbarItem itemIndex={0} aria-haspopup="tree">
          <ButtonPill aria-haspopup="dialog" aria-label="label text">
            Test
          </ButtonPill>
        </AriaToolbarItem>
      );

      expect(container.find('ButtonPill').props()).toEqual({
        children: 'Test',
        'aria-haspopup': 'dialog', // item has "tree", but children has "dialog", children's props takes precedent
        'aria-label': 'label text',
        onFocus: expect.any(Function),
        onKeyDown: expect.any(Function),
        onKeyUp: undefined,
        onPress: expect.any(Function),
        tabIndex: 0,
        useNativeKeyDown: true,
      });
    });
  });

  // keyboard interaction is tested in AriaToolbar.unit.test.tsx
});
