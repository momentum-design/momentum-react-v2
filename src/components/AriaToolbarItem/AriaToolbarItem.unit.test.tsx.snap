// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AriaToolbarItem /> snapshot should match snapshot 1`] = `
<TestComponent>
  <ForwardRef
    itemIndex={0}
  >
    <ButtonPill
      onFocus={[Function]}
      onKeyDown={[Function]}
      onPress={[Function]}
      tabIndex={0}
      useNativeKeyDown={true}
    >
      <ButtonSimple
        aria-disabled={false}
        className="md-button-pill-wrapper"
        data-color="primary"
        data-disabled={false}
        data-ghost={false}
        data-grown={false}
        data-inverted={false}
        data-only-triggers-tooltip={false}
        data-outline={false}
        data-shallow-disabled={false}
        data-size={40}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onPress={[Function]}
        tabIndex={0}
        useNativeKeyDown={true}
      >
        <FocusRing>
          <FocusRing
            focusClass="md-focus-ring-wrapper children"
            focusRingClass="md-focus-ring-wrapper children"
          >
            <button
              className="md-button-pill-wrapper md-button-simple-wrapper"
              data-color="primary"
              data-disabled={false}
              data-ghost={false}
              data-grown={false}
              data-inverted={false}
              data-only-triggers-tooltip={false}
              data-outline={false}
              data-shallow-disabled={false}
              data-size={40}
              onBlur={[Function]}
              onClick={[Function]}
              onDragStart={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyUp={[Function]}
              onMouseDown={[Function]}
              onMouseEnter={[Function]}
              onMouseLeave={[Function]}
              onMouseUp={[Function]}
              onTouchCancel={[Function]}
              onTouchEnd={[Function]}
              onTouchMove={[Function]}
              onTouchStart={[Function]}
              type="button"
            >
              <span>
                Test
              </span>
            </button>
          </FocusRing>
        </FocusRing>
      </ButtonSimple>
    </ButtonPill>
  </ForwardRef>
</TestComponent>
`;
